lib/TemplatedPathPlugin.js

```diff
@@ -75,130 +90,187 @@ const getReplacer = (value, allowEmpty) => {
+	// [filebase] - data.basename (file.map.js)
+	if (data.chunk) {
+		const chunk = data.chunk;
+
+		const id = chunk.id;
+		const name = chunk.name || chunk.id;
+
+		const chunkHash = chunk.renderedHash || chunk.hash;
+		// @ts-ignore
+		const chunkHashWithLength = chunk.hashWithLength || undefined;
+
+		const contentHashType = data.contentHashType;
+		const contentHash =
+			data.contentHash ||
+			(contentHashType &&
+				chunk.contentHash &&
+				chunk.contentHash[contentHashType]);
+		const contentHashWithLength =
+			data.contentHashWithLength ||
+			// @ts-ignore
+			(chunk.contentHashWithLength &&
+				// @ts-ignore
+				chunk.contentHashWithLength[contentHashType]) ||
+			undefined;
+
+		if (typeof path === "function") {
+			path = path(data);
+		}
+
+		if (
+			data.noChunkHash &&
+			(REGEXP_CHUNKHASH_FOR_TEST.test(path) ||
+				REGEXP_CONTENTHASH_FOR_TEST.test(path))
+		) {
+			throw new Error(
+				`Cannot use [chunkhash] or [contenthash] for chunk in '${path}' (use [hash] instead)`
+			);
+		}
+
+		return (
+			path
+				.replace(REGEXP_ID, replacer(id))
+				.replace(REGEXP_NAME, replacer(name))
+				.replace(
+					REGEXP_HASH,
+					hashLength(replacer(data.hash), data.hashWithLength)
+				)
+				.replace(
+					REGEXP_CHUNKHASH,
+					hashLength(replacer(chunkHash), chunkHashWithLength)
+				)
+				.replace(
+					REGEXP_CONTENTHASH,
+					hashLength(replacer(contentHash), contentHashWithLength)
+				)
+				// query is optional, it's OK if it's in a path
+				// but there's nothing to replace it with
+				.replace(REGEXP_QUERY, replacer(data.query, true))
+				.replace(REGEXP_FILE, replacer(data.filename))
+				.replace(REGEXP_FILEBASE, replacer(data.basename))
+			// TODO
+			// .replace(REGEXP_EXT, replacer(extensions[type]))
 		);
 	}
 
-	return (
-		path
-			.replace(
-				REGEXP_HASH,
-				withHashLength(getReplacer(data.hash), data.hashWithLength)
-			)
-			.replace(
-				REGEXP_CHUNKHASH,
-				withHashLength(getReplacer(chunkHash), chunkHashWithLength)
-			)
-			.replace(
-				REGEXP_CONTENTHASH,
-				withHashLength(getReplacer(contentHash), contentHashWithLength)
-			)
-			.replace(
-				REGEXP_MODULEHASH,
-				withHashLength(getReplacer(moduleHash), moduleHashWithLength)
-			)
-			.replace(REGEXP_ID, getReplacer(chunkId))
-			.replace(REGEXP_MODULEID, getReplacer(prepareId(moduleId)))
-			.replace(REGEXP_NAME, getReplacer(chunkName))
-			.replace(REGEXP_FILE, getReplacer(data.filename))
-			.replace(REGEXP_FILEBASE, getReplacer(data.basename))
-			// query is optional, it's OK if it's in a path but there's nothing to replace it with
-			.replace(REGEXP_QUERY, getReplacer(data.query, true))
-	);
+	// Module Context
+	//
+	// Placeholders
+	//
+	// [id] - module.id (2.png)
+	// [ext] - extname(module.resource || data.filenam) (file.{png, svg, ...})
+	// [hash] - module.hash (6237543873.png)
+	// [name] - file.replace(extname, '') (file)
+	// [file] - basename(module.resource || data.filename) (file.png)
```

Comment:
`module.resource` is not a probably of every module and can be used here. (only the comment is incorrect)

Reply:
You mean `module.resource` "**can't** be used" instead of can? I noticed this for `wasm` modules (tests) for e.g, but wasn't 💯 why and if `module.resource` can be `undefined`, better to explicitly rely on `data.filename` in general imho and treat it as the 'single source of truth' here (makes handling slightly easier aswell)

Reply:
ah yes it can't be used.